File: examples/tutorial_ping_pong/ping_actor.lua
function receive(message)
  if(message.type == "init" or (message.type == "wakeup" and message.wakeup_id == "ping_trigger")) then

    if(message.type == "init") then
      print("Start ping actor")
      sequence_number = 0  
    else
      sequence_number = sequence_number + 1
    end

    publish(Publication.new(
      "type", "ping",
      "sequence_number", sequence_number
    ))
    print("ping sent: "..sequence_number) 

    enqueue_wakeup(5000, "ping_trigger")
  elseif(message.type == "exit") then
    print("Stop ping actor")
  elseif(message.type == "pong") then
    print("pong received: "..message.sequence_number)
  end
end

File: examples/tutorial_ping_pong/ping_actor.js
function receive(message, state) {
  if(message.type == "init") {
    log("Start JS Ping Actor")
    state.sequence_number = 0
    send_ping(state)
    enqueue_wakeup(5000, "ping_trigger")
  } else if(message.type == "wakeup" && message.wakeup_id == "ping_trigger") {
    send_ping(state)
    enqueue_wakeup(5000, "ping_trigger") 
  } else if(message.type == "exit") {
    log("Stop JS Ping Actor")
  } else if(message.type == "pong") {
    log("Pong Received: "+message.sequence_number)
  }
}

function send_ping(state) {
  let sequence_number = state.sequence_number++
  log("Send Ping: "+sequence_number)
  publish(new Message({
    type: "ping",
    sequence_number: sequence_number
  }))
}
